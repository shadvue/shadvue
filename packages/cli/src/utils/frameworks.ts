export const FRAMEWORKS = {
  vite: {
    name: 'vite',
    label: 'Vite',
    links: {
      installation: 'https://shadcn-vue.com/docs/installation/vite',
      tailwind: 'https://tailwindcss.com/docs/guides/vite',
    },
  },
  nuxt: {
    name: 'nuxt',
    label: 'Nuxt',
    links: {
      installation: 'https://shadcn-vue.com/docs/installation/nuxt',
      tailwind: 'https://tailwindcss.com/docs/guides/nuxtjs',
    },
  },
  astro: {
    name: 'astro',
    label: 'Astro',
    links: {
      installation: 'https://shadcn-vue.com/docs/installation/astro',
      tailwind: 'https://tailwindcss.com/docs/guides/astro',
    },
  },
  laravel: {
    name: 'laravel',
    label: 'Laravel',
    links: {
      installation: 'https://shadcn-vue.com/docs/installation/laravel',
      tailwind: 'https://tailwindcss.com/docs/guides/laravel',
    },
  },
} as const

export type Framework = (typeof FRAMEWORKS)[keyof typeof FRAMEWORKS]
