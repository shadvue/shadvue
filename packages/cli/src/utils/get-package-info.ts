import type { PackageJson } from 'type-fest'
import fs from 'fs-extra'
import path from 'pathe'

export function getPackageInfo(
  cwd: string = '',
  shouldThrow: boolean = true,
): PackageJson | null {
  const packageJsonPath = path.join(cwd, 'package.json')

  return fs.readJSONSync(packageJsonPath, {
    throws: shouldThrow,
  }) as PackageJson
}
