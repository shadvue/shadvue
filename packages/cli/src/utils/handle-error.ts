import { consola } from 'consola'

export function handleError(error: unknown) {
  consola.log('this is error: ', error)
  if (typeof error === 'string') {
    consola.error(error)
    process.exit(1)
  }

  if (error instanceof Error) {
    consola.error(error.message)
    process.exit(1)
  }

  consola.error('Something went wrong. Please try again.')
  process.exit(1)
}
