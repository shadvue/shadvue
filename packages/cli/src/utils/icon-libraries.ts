export const ICON_LIBRARIES = {
  lucide: {
    name: 'lucide-vue-next',
    package: 'lucide-vue-next',
    import: 'lucide-vue-next',
  },
  radix: {
    name: '@radix-icons/vue',
    package: '@radix-icons/vue',
    import: '@radix-icons/vue',
  },
}
