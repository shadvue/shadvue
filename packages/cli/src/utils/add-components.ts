import type { Config, configSchema } from '@/src/utils/get-config'
import type { z } from 'zod'
import { getProjectTailwindVersionFromConfig } from '@/src/utils/get-project-info'
import { handleError } from '@/src/utils/handle-error'
import { logger } from '@/src/utils/logger'
import { registryResolveItemsTree } from '@/src/utils/registry'
import { spinner } from '@/src/utils/spinner'
import { updateCssVars } from '@/src/utils/updaters/update-css-vars'
import { updateDependencies } from '@/src/utils/updaters/update-dependencies'
import { updateFiles } from '@/src/utils/updaters/update-files'
import { updateTailwindConfig } from '@/src/utils/updaters/update-tailwind-config'

export async function addComponents(
  components: string[],
  config: Config,
  options: {
    overwrite?: boolean
    silent?: boolean
    isNewProject?: boolean
  },
) {
  options = {
    overwrite: false,
    silent: false,
    isNewProject: false,
    ...options,
  }

  // TODO: Add workspace component
  return await addProjectComponents(components, config, options)
}

async function addProjectComponents(
  components: string[],
  config: z.infer<typeof configSchema>,
  options: {
    overwrite?: boolean
    silent?: boolean
    isNewProject?: boolean
  },
) {
  const registrySpinner = spinner(`Checking registry.`, {
    silent: options.silent,
  })?.start()
  const tree = await registryResolveItemsTree(components, config)
  if (!tree) {
    registrySpinner?.fail()
    return handleError(new Error('Failed to fetch components from registry.'))
  }
  registrySpinner?.succeed()

  const tailwindVersion = await getProjectTailwindVersionFromConfig(config)

  await updateTailwindConfig(tree.tailwind?.config, config, {
    silent: options.silent,
    tailwindVersion,
  })
  await updateCssVars(tree.cssVars, config, {
    cleanupDefaultNextStyles: options.isNewProject,
    silent: options.silent,
    tailwindVersion,
    tailwindConfig: tree.tailwind?.config,
  })

  await updateDependencies(tree.dependencies, config, {
    silent: options.silent,
  })
  await updateFiles(tree.files, config, {
    overwrite: options.overwrite,
    silent: options.silent,
  })

  if (tree.docs) {
    logger.info(tree.docs)
  }
}
