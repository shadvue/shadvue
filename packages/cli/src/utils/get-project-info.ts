import type { Framework } from '@/src/utils/frameworks'
import type {
  Config,
  RawConfig,
} from '@/src/utils/get-config'
import { FRAMEWORKS } from '@/src/utils/frameworks'
import {
  getConfig,
  getTSConfig,
  resolveConfigPaths,
} from '@/src/utils/get-config'
import { getPackageInfo } from '@/src/utils/get-package-info'
import fs from 'fs-extra'
import { parseTsconfig } from 'get-tsconfig'
import path from 'pathe'
import { glob } from 'tinyglobby'
import { z } from 'zod'

export interface ProjectInfo {
  framework: Framework
  typescript: boolean
  tailwindConfigFile: string | null
  tailwindCssFile: string | null
  aliasPrefix: string | null
}

const PROJECT_SHARED_IGNORE = [
  '**/node_modules/**',
  '.nuxt',
  'public',
  'dist',
  'build',
]

const TS_CONFIG_SCHEMA = z.object({
  compilerOptions: z.object({
    paths: z.record(z.string().or(z.array(z.string()))),
  }),
})

export async function getProjectInfo(cwd: string): Promise<ProjectInfo | null> {
  const [
    configFiles,
    typescript,
    tailwindConfigFile,
    tailwindCssFile,
    aliasPrefix,
    packageJson,
  ] = await Promise.all([
    glob('**/{nuxt,vite,astro}.config.*|composer.json', {
      cwd,
      deep: 3,
      ignore: PROJECT_SHARED_IGNORE,
    }),
    isTypeScriptProject(cwd),
    getTailwindConfigFile(cwd),
    getTailwindCssFile(cwd),
    getTsConfigAliasPrefix(cwd),
    getPackageInfo(cwd, false),
  ])

  const type: ProjectInfo = {
    framework: FRAMEWORKS.vite, // TODO: Maybe add a manual installation
    typescript,
    tailwindConfigFile,
    tailwindCssFile,
    aliasPrefix,
  }

  // Nuxt.
  if (configFiles.find(file => file.startsWith('nuxt.config.'))?.length) {
    type.framework = FRAMEWORKS.nuxt
    return type
  }

  // Astro.
  if (configFiles.find(file => file.startsWith('astro.config.'))?.length) {
    type.framework = FRAMEWORKS.astro
    return type
  }

  // Laravel.
  if (configFiles.find(file => file.startsWith('composer.json'))?.length) {
    type.framework = FRAMEWORKS.laravel
    return type
  }

  // Vite.
  // We'll assume that it got caught by the Remix check above.
  if (configFiles.find(file => file.startsWith('vite.config.'))?.length) {
    type.framework = FRAMEWORKS.vite
    return type
  }

  return type
}

export async function getTailwindCssFile(cwd: string) {
  const files = await glob(['**/*.css', '**/*.scss'], {
    cwd,
    deep: 5,
    ignore: PROJECT_SHARED_IGNORE,
  })

  if (!files.length) {
    return null
  }

  for (const file of files) {
    const contents = await fs.readFile(path.resolve(cwd, file), 'utf8')
    // Assume that if the file contains `@tailwind base` it's the main css file.
    if (contents.includes('@tailwind base')) {
      return file
    }
  }

  return null
}

export async function getTailwindConfigFile(cwd: string) {
  const files = await glob('tailwind.config.*', {
    cwd,
    deep: 3,
    ignore: PROJECT_SHARED_IGNORE,
  })

  if (!files.length) {
    return null
  }

  return files[0]
}

export async function getTsConfigAliasPrefix(cwd: string) {
  const isTypescript = await isTypeScriptProject(cwd)
  const tsconfigType = isTypescript ? 'tsconfig.json' : 'jsconfig.json'

  const tsConfig = getTSConfig(cwd, tsconfigType)
  const parsedTsConfig = parseTsconfig(tsConfig.path)

  const aliasPaths = parsedTsConfig.compilerOptions?.paths ?? {}

  // This assume that the first alias is the prefix.
  for (const [alias, paths] of Object.entries(aliasPaths)) {
    if (
      paths.includes('./*')
      || paths.includes('./src/*')
      || paths.includes('./app/*')
      || paths.includes('./resources/js/*') // Laravel.
    ) {
      const cleanAlias = alias.replace(/\/\*$/, '') ?? null
      // handle Nuxt
      return cleanAlias === '#build' ? '@' : cleanAlias
    }
  }

  // Use the first alias as the prefix.
  return Object.keys(aliasPaths)?.[0]?.replace(/\/\*$/, '') ?? null
}

export async function isTypeScriptProject(cwd: string) {
  const files = await glob('tsconfig.*', {
    cwd,
    deep: 1,
    ignore: PROJECT_SHARED_IGNORE,
  })

  return files.length > 0
}

export async function getTsConfig(cwd: string) {
  for (const fallback of [
    'tsconfig.json',
    'tsconfig.web.json',
    'tsconfig.app.json',
  ]) {
    const filePath = path.resolve(cwd, fallback)
    if (!(await fs.pathExists(filePath))) {
      continue
    }

    // We can't use fs.readJSON because it doesn't support comments.
    const contents = await fs.readFile(filePath, 'utf8')
    const cleanedContents = contents.replace(/\/\*\s*\*\//g, '')
    const result = TS_CONFIG_SCHEMA.safeParse(JSON.parse(cleanedContents))

    if (result.error) {
      continue
    }

    return result.data
  }

  return null
}

export async function getProjectConfig(
  cwd: string,
  defaultProjectInfo: ProjectInfo | null = null,
): Promise<Config | null> {
  // Check for existing component config.
  const [existingConfig, projectInfo] = await Promise.all([
    getConfig(cwd),
    !defaultProjectInfo
      ? getProjectInfo(cwd)
      : Promise.resolve(defaultProjectInfo),
  ])

  if (existingConfig) {
    return existingConfig
  }

  if (
    !projectInfo
    || !projectInfo.tailwindConfigFile
    || !projectInfo.tailwindCssFile
  ) {
    return null
  }

  const config: RawConfig = {
    $schema: 'https://shadcn-vue.com/schema.json',
    typescript: true,
    style: 'new-york',
    tailwind: {
      config: projectInfo.tailwindConfigFile,
      baseColor: 'zinc',
      css: projectInfo.tailwindCssFile,
      cssVariables: true,
      prefix: '',
    },
    iconLibrary: 'lucide',
    aliases: {
      components: `${projectInfo.aliasPrefix}/components`,
      ui: `${projectInfo.aliasPrefix}/components/ui`,
      composables: `${projectInfo.aliasPrefix}/composables`,
      lib: `${projectInfo.aliasPrefix}/lib`,
      utils: `${projectInfo.aliasPrefix}/lib/utils`,
    },
  }

  return await resolveConfigPaths(cwd, config)
}
