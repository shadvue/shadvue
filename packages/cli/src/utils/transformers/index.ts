import type { Config } from '@/src/utils/get-config'
import type { registryBaseColorSchema } from '@/src/utils/registry/schema'
import type * as z from 'zod'
import { getRegistryIcons } from '@/src/utils/registry'
import { transformCssVars } from '@/src/utils/transformers/transform-css-vars'
import { transformImport } from '@/src/utils/transformers/transform-import'
import { transformSFC } from '@/src/utils/transformers/transform-sfc'
import { transformTwPrefix } from '@/src/utils/transformers/transform-tw-prefix'
import { transform as metaTransform } from 'vue-metamorph'
import { transformIcons } from './transform-icons'

export interface TransformOpts {
  filename: string
  raw: string
  config: Config
  baseColor?: z.infer<typeof registryBaseColorSchema>
}

export async function transform(opts: TransformOpts) {
  const source = await transformSFC(opts)

  const registryIcons = await getRegistryIcons()

  return metaTransform(source, opts.filename, [
    transformImport(opts),
    transformCssVars(opts),
    transformTwPrefix(opts),
    transformIcons(opts, registryIcons),
  ]).code
}
