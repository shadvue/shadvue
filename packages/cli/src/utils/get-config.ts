import { cosmiconfig } from 'cosmiconfig'
import { getTsconfig } from 'get-tsconfig'
import path from 'pathe'
import { z } from 'zod'
import { getProjectInfo } from '@/src/utils/get-project-info'
import { resolveImport } from '@/src/utils/resolve-import'
import { highlighter } from './highlighter'

/** @deprecated */
export const TAILWIND_CSS_PATH = {
  nuxt: 'assets/css/tailwind.css',
  vite: 'src/assets/index.css',
  laravel: 'resources/css/app.css',
  astro: 'src/styles/globals.css',
}

export const DEFAULT_STYLE = 'default'
export const DEFAULT_COMPONENTS = '@/components'
export const DEFAULT_UTILS = '@/lib/utils'
export const DEFAULT_TAILWIND_CSS = TAILWIND_CSS_PATH.nuxt // decide to go with Nuxt's as default
export const DEFAULT_TAILWIND_CONFIG = 'tailwind.config.js'
export const DEFAULT_TAILWIND_BASE_COLOR = 'slate'
export const DEFAULT_TYPESCRIPT_CONFIG = './tsconfig.json'

// zernonia: replaced this from `c12` because it cause error with `components` folder in Nuxt.
// TODO: Figure out if we want to support all cosmiconfig formats.
// A simple components.json file would be nice.
const explorer = cosmiconfig('components', {
  searchPlaces: ['components.json'],
})

export const rawConfigSchema = z
  .object({
    $schema: z.string().optional(),
    style: z.string(),
    typescript: z.boolean().default(true),
    tailwind: z.object({
      config: z.string().optional(),
      css: z.string(),
      baseColor: z.string(),
      cssVariables: z.boolean().default(true),
      prefix: z.string().default('').optional(),
    }),
    aliases: z.object({
      components: z.string(),
      composables: z.string().optional(),
      utils: z.string(),
      ui: z.string().optional(),
      lib: z.string().optional(),
    }),
    iconLibrary: z.string().optional(),
  })
  .strict()

export type RawConfig = z.infer<typeof rawConfigSchema>

export const configSchema = rawConfigSchema.extend({
  resolvedPaths: z.object({
    cwd: z.string(),
    tailwindConfig: z.string(),
    tailwindCss: z.string(),
    utils: z.string(),
    components: z.string(),
    composables: z.string(),
    lib: z.string(),
    ui: z.string(),
  }),
})

export type Config = z.infer<typeof configSchema>

export async function getConfig(cwd: string) {
  const config = await getRawConfig(cwd)

  if (!config) {
    return null
  }

  // Set default icon library if not provided.
  if (!config.iconLibrary) {
    config.iconLibrary = config.style === 'new-york' ? 'radix' : 'lucide'
  }

  return await resolveConfigPaths(cwd, config)
}

export function getTSConfig(cwd: string, tsconfigName: 'tsconfig.json' | 'jsconfig.json') {
  const parsedConfig = getTsconfig(path.resolve(cwd, 'package.json'), tsconfigName)
  if (parsedConfig === null) {
    throw new Error(
      `Failed to find ${highlighter.info(tsconfigName)}`,
    )
  }

  return parsedConfig
}

export async function resolveConfigPaths(cwd: string, config: RawConfig) {
  // Read tsconfig.json.
  const tsconfigType = config.typescript ? 'tsconfig.json' : 'jsconfig.json'
  const tsConfig = getTSConfig(cwd, tsconfigType)

  return configSchema.parse({
    ...config,
    resolvedPaths: {
      cwd,
      tailwindConfig: config.tailwind.config
        ? path.resolve(cwd, config.tailwind.config)
        : '',
      tailwindCss: path.resolve(cwd, config.tailwind.css),
      utils: await resolveImport(config.aliases.utils, tsConfig),
      components: await resolveImport(config.aliases.components, tsConfig),
      ui: config.aliases.ui
        ? await resolveImport(config.aliases.ui, tsConfig)
        : path.resolve(
            (await resolveImport(config.aliases.components, tsConfig))
            ?? cwd,
            'ui',
          ),
      // TODO: Make this configurable.
      // For now, we assume the lib and hooks directories are one level up from the components directory.
      lib: config.aliases.lib
        ? await resolveImport(config.aliases.lib, tsConfig)
        : path.resolve(
            (await resolveImport(config.aliases.utils, tsConfig)) ?? cwd,
            '..',
          ),
      composables: config.aliases.composables
        ? await resolveImport(config.aliases.composables, tsConfig)
        : path.resolve(
            (await resolveImport(config.aliases.components, tsConfig))
            ?? cwd,
            '..',
            'composables',
          ),
    },
  })
}

export async function getRawConfig(cwd: string): Promise<RawConfig | null> {
  try {
    const configResult = await explorer.search(cwd)
    if (!configResult) {
      return null
    }

    return rawConfigSchema.parse(configResult.config)
  }
  catch (error) {
    throw new Error(`Invalid configuration found in ${cwd}/components.json.`)
  }
}

// TODO: Cache this call.
export async function getTargetStyleFromConfig(cwd: string, fallback: string) {
  const projectInfo = await getProjectInfo(cwd)
  return projectInfo?.tailwindVersion === 'v4' ? 'new-york-v4' : fallback
}
