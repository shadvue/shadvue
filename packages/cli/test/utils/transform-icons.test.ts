import { describe, expect, it } from 'vitest'
import { transform } from '../../src/utils/transformers'

describe('transformIcons', () => {
  it('transforms radix icons', async () => {
    const result = await transform({
      filename: 'app.vue',
      raw: `<script lang="ts" setup>
      import { Check } from 'lucide-vue-next'
      import { Primitive } from 'reka-ui'
      </script>

      <template>
        <Check />
        <Primitive />
      </template>
      `,
      config: {
        iconLibrary: 'radix',
      },
    })
    expect(result).toMatchSnapshot()
  })

  it('does not transform lucide icons', async () => {
    const result = await transform({
      filename: 'app.vue',
      raw: `<script lang="ts" setup>
      import { Check } from 'lucide-vue-next'
      import { Primitive } from 'reka-ui'
      </script>

      <template>
        <Check />
        <Primitive />
      </template>
      `,
      config: {
        iconLibrary: 'lucide',
      },
    })
    expect(result).toMatchSnapshot()
  })

  it('does nothing', async () => {
    const result = await transform({
      filename: 'app.vue',
      raw: `<script lang="ts" setup>
      import { Check } from 'lucide-vue-next'
      import { Primitive } from 'reka-ui'
      </script>

      <template>
        <Check />
        <Primitive />
      </template>
      `,
      config: {
        style: 'new-york',
        tailwind: {
          config: 'tailwind.config.js',
          css: 'src/assets/index.css',
          baseColor: 'zinc',
          cssVariables: true,
        },
        aliases: {
          utils: '@/utils',
          components: '@/components',
        },
      },
    })
    expect(result).toMatchSnapshot()
  })
})
