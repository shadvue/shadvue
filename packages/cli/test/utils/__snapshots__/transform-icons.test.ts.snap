// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`transformIcons > does not transform lucide icons 1`] = `
"<script setup>
import { Check } from 'lucide-vue-next';
import { Primitive } from 'reka-ui';
</script>

<template>
  <Check />
  <Primitive />
</template>
"
`;

exports[`transformIcons > does nothing 1`] = `
"<script setup>
import { Check } from 'lucide-vue-next';
import { Primitive } from 'reka-ui';
</script>

<template>
  <Check />
  <Primitive />
</template>
"
`;

exports[`transformIcons > transforms radix icons 1`] = `
"<script setup>
import { CheckIcon } from '@radix-icons/vue';
import { Primitive } from 'reka-ui';
</script>

<template>
  <CheckIcon />
  <Primitive />
</template>
"
`;
